src/cr/cube/cube_slice.py

```diff
@@ -309,6 +329,29 @@ def _apply_pruning_mask(self, res, prune):
 
         return np.ma.masked_array(res, mask=array.mask)
 
+    def _calculate_std_res(self, counts, total, colsum, rowsum):
+        has_mr_or_ca = set(self.dim_types) & DT.ARRAY_TYPES
+        if has_mr_or_ca:
+            if self.mr_dim_ind == 0:
+                total = total[:, np.newaxis]
+                rowsum = rowsum[:, np.newaxis]
+
+            expected = rowsum * colsum / total
+            variance = (
+                rowsum * colsum * (total - rowsum) * (total - colsum) /
+                total ** 3
+            )
+            res = (counts - expected) / np.sqrt(variance)
+        else:
+            expected_counts = expected_freq(counts)
+            residuals = counts - expected_counts
+            variance = (
+                np.outer(rowsum, colsum) *
+                np.outer(total - rowsum, total - colsum) / total ** 3
+            )
+            res = residuals / np.sqrt(variance)
+        return res
```

Comment:
I think this would be better as:

```python
    def _calculate_std_res(self, counts, total, colsum, rowsum):
        """Return ndarray containing standard residuals.

        The shape of the return value is the same as that of *counts*.
        """
        has_mr_or_ca = set(self.dim_types) & DT.ARRAY_TYPES
        if has_mr_or_ca:
            if self.mr_dim_ind == 0:
                total = total[:, np.newaxis]
                rowsum = rowsum[:, np.newaxis]
            expected_counts = rowsum * colsum / total
            variance = (
                rowsum * colsum * (total - rowsum) * (total - colsum) /
                total ** 3
            )
            return (counts - expected_counts) / np.sqrt(variance)

        expected_counts = expected_freq(counts)
        residuals = counts - expected_counts
        variance = (
            np.outer(rowsum, colsum) *
            np.outer(total - rowsum, total - colsum) / total ** 3
        )
        return residuals / np.sqrt(variance)
```

and even better as:

```python
    def _calculate_std_res(self, counts, total, colsum, rowsum):
        """Return ndarray containing standard residuals.

        The shape of the return value is the same as that of *counts*.
        """
        if set(self.dim_types) & DT.ARRAY_TYPES:  # ---has-mr-or-ca---
            return self._array_type_std_res(counts, total, colsum, rowsum)
        return self._cat_type_std_res(counts, total, colsum, rowsum)
        # ---or maybe this is called _scalar_type_std_res()---

    def _array_type_std_res(self, counts, total, colsum, rowsum):
        """Return ndarray containing standard residuals for array values.

        The shape of the return value is the same as that of *counts*.
        Array variables require special processing because ...
        """
        if self.mr_dim_ind == 0:
            # --helpful comment explaining to me why this preliminary
            # --(inflation?) operation is required and what this special case
            # --is.
            total = total[:, np.newaxis]
            rowsum = rowsum[:, np.newaxis]
        expected_counts = rowsum * colsum / total
        variance = (
            rowsum * colsum * (total - rowsum) * (total - colsum) /
            total ** 3
        )
        return (counts - expected_counts) / np.sqrt(variance)

    def _scalar_type_std_res(self, counts, total, colsum, rowsum):
        """Return ndarray ..."""
        expected_counts = expected_freq(counts)
        residuals = counts - expected_counts
        variance = (
            np.outer(rowsum, colsum) *
            np.outer(total - rowsum, total - colsum) / total ** 3
        )
        return residuals / np.sqrt(variance)
```

Which is all staged up for later extraction into separate subclasses, like maybe ScalarSlice and ArraySlice. In any case, each method is now more compact and simpler to reason about, especially the `_scalar_type_std_res()` which is now straight-through (no if statements), and very importantly, they can now be thoroughly unit tested.

The first and the third are easily unit tested. The second one would require a parameterized fixture to handle the multiple cases, but is still well within range of unit testing.

I'm not sure if the right name is `_scalar_type_std_res` or perhaps `_cat_type_std_res`, but in either case it's an opportunity to articulate in the code a fundamental distinction in processing cases that I bet shows up elsewhere as well (and can then be signaled as similar by using the same prefix name array or scalar). These distinctions have a way of growing into subclasses that gather those differences in processing (polymorphishm).

Note that these methods need docstrings that clearly articulate their contract, the same as any other method. You don't need those now because you wrote them, but I need it to understand the behavior and you need them later to remember. Also they have a way of letting you know when you haven't got the contract right because it's too hard to explain in a concise docstring.

Reply:
Done and unit tested (still only partially)